pysnark/runtime.py

```diff
@@ -204,97 +204,203 @@ def __ge__(self, other): return (self-other).check_positive()
-            if self.value % other != 0:
-                raise ValueError(str(self.value) + " is not properly divisible by " + str(other))
-            return LinComb(self.value/other, self.lc*backend.fieldinverse(other))
-        else:
-            return NotImplemented
+
+        return NotImplemented
          
     def __floordiv__(self, other):
-        """ Division with rounding """
-        return self.__divmod__(other)[0]
+        """
+        Divides a LinComb with an integer or another LinComb using floor division
+        Costs 0 constraints to divide with an integer
+        Costs 1 constraint to divide with a LinComb
+        """
+        res = self.__divmod__(other)
+        if res is NotImplemented:
+            return NotImplemented
+        return res[0]
 
     def __mod__(self, other):
-        if other&(other-1)==0:
-            # this is faster for powers of two
-            return LinComb.from_bits(self.to_bits()[:other.bit_length()-1])
-        
-        return self.__divmod__(other)[1]
+        """
+        Returns the remainder of a LinComb divided with an integer or another LinComb
+        Costs 0 constraints to divide modulo an integer
+        Costs 1 constraint to divide modulo a LinComb
+        """
+        res = self.__divmod__(other)
+        if res is NotImplemented:
+            return NotImplemented
+        return res[1]
         
     def __divmod__(self, divisor):
-        """ Division by public value """
-        
-        if not is_base_value(divisor): return NotImplemented
-        
-        if divisor==0: raise ValueError("division by zero")
-        
-        quo = PrivVal(self.value//divisor)
-        rem = PrivVal(self.value-quo.value*divisor)
- 
-        rem.assert_positive(divisor.bit_length())
-        if divisor&(divisor-1)!=0: rem.assert_lt(divisor) # not needed for powers of two
-        quo.assert_positive()
-        (self-divisor*quo-rem).assert_zero()
-        
-        return (quo,rem)
+        """
+        Divides a LinComb with an integer or another LinComb and returns the quotient and the remainder
+        Costs 0 constraints to divide with an integer
+        Costs 3 constraints to divide with a LinComb
+        """
+        if isinstance(divisor, int):
+            if divisor == 0:
+                raise ValueError("Division by zero")
+            quo = ConstVal(self.value // divisor)
+            rem = ConstVal(self.value % divisor)
+            return (quo, rem)
+
+        if isinstance(divisor, LinComb):
+            if divisor.value == 0:
+                raise ValueError("Division by zero")
+            quo = PrivVal(self.value // divisor.value)
+            res = quo * divisor
+            rem = PrivVal(self.value - res.value)
+
+            # TODO: Prevent field overflow by a malicious prover
+            # Add a constraint ensuring quo * divisor + rem < backend.get_modulus()
+            # The check above works over the integers, but not in a finite field
```

Comment:
What problem do you see here? Both quo and rem are bounded by the use of assert_lt and assert_positive, right?

Reply:
Let's say our field has size 10 and we're computing `(5).__divmod__(3)`
The correct `quo` is 1 and the correct `rem` is 2.

However, we could choose quo such that the result overflows the field. Then, we can find a rem that works. For example:
If `quo = 3`, let `rem = 6`. Then `divisor * quo + rem = 3 * 3 + 6 = 15 = 5 (mod 10)`.
If `quo = 4`, let `rem = 3`. Then `divisor * quo + rem = 3 * 4 + 3 = 15 = 5 (mod 10)`.
If `quo = 5`, let `rem = 0`. Then `divisor * quo + rem = 3 * 5 + 0 = 15 = 5 (mod 10)`.

Thus, a malicious prover could choose any `rem` to satisfy the constraint.

Is this a problem?

Reply:
Yes, but an assert_lt is done on quo, which can only pass if 0<=quo<2**bit_length?

Reply:
Did you mean `assert_positive`? I see. Then we'd have to prevent bitlength being set too high.
I guess this is something that could go in the docs instead as a warning/limitation.

Reply:
Ah yes, I meant assert_positive. You mean that bitlength should be set smaller than the modulus? Yes, that is definitely important :)